java_src/foam/dao/index/TreeNode.java

```diff
@@ -0,0 +1,352 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package foam.dao.index;
+
+import foam.core.FObject;
+import foam.core.PropertyInfo;
+
+
+public class TreeNode {
+  
+  protected Index tail;
```

Comment:
Why does the TreeNode need to have the Index. You'll have 100's of millions of TreeNodes, so any extra size adds up.